drivers: gpio_rpi_pico: Extend the driver API

ajf58 opened the pull request:
Edit:

The original goal of this PR was to add just the `gpio_get_config`. This PR has now widened to include a lot more functionality, extending the both the existing APIs to be more feature-complete, and adding new APIs that were missing.

Tested locally using a physical test fixture supporting the GPIO loopback.

ajf58 committed:
drivers: gpio_rpi_pico: Add gpio_get_config API

Implement the `gpio_get_config`

N.b. adding this API results in a new test failure in
`test_gpio_config_trigger`. This suggests that there is some kind of
dependency between this and the now-enabled `pin_get_config` test cases.
Note that this adds a read-only API, it is unlikely to be the cause of
the failure.

Signed-off-by: Andrew Featherstone <andrew.featherstone@gmail.com>

ajf58 committed:
drivers: gpio_rpi_pico: Add gpio_get_pending_int API

Implement `gpio_get_pending_int`.

Signed-off-by: Andrew Featherstone <andrew.featherstone@gmail.com>

ajf58 commented:
cc: @drensber

soburi commented on drivers/gpio/gpio_rpi_pico.c:
```diff
@@ -80,6 +80,29 @@ static int gpio_rpi_configure(const struct device *dev,
 }
 
+#ifdef CONFIG_GPIO_GET_CONFIG
+static int gpio_rpi_get_config(const struct device *dev, gpio_pin_t pin, gpio_flags_t *flags)
```
It is better to handle the flags that are handled by `gpio_rpi_configure` symmetrically.

ThreeEights replied:
I have to agree:  what you get back from `gpio_rpi_get_config` should be as close as possible to what you put into `gpio_rpi_configure`, combined with settings from pinctrl.

ajf58 replied:
I don't disagree, however.... we didn't have this API call at all, and this PR gives _some_ correct functionality. In trying to have _complete_ functionality it takes a lot more work (80/20 rule at play here).

I'm working on this, but it's a grind.

soburi replied:
It will inevitably take time to improve quality.
Fortunately, we're not in a rush on this topic.

soburi commented on tests/drivers/gpio/gpio_basic_api/boards/rpi_pico2_rp2350a_m33.conf:
```diff
@@ -0,0 +1 @@
+CONFIG_SKIP_PULL_TEST=y
```
@henrikbrixandersen 

Let me confirm one thing.
Currently, Pico2 has an errata that causes incorrect behavior when GPIO_INPUT and GPIO_PULL_DOWN are specified at the same time. For this reason, it specifies SKIP_PULL here to avoid testing this case, but is this solution okay?

henrikbrixandersen replied:
Yes.

soburi replied:
Thank you for your reply. I'll move forward with this.

ajf58 committed:
drivers: gpio_rpi_pico: Add gpio_port_get_direction API

THe driver didn't implement this API, so add it.

Signed-off-by: Andrew Featherstone <andrew.featherstone@gmail.com>

ajf58 committed:
drivers: gpio_rpi_pico: Support GPIO_DISCONNECTED flag

Out of reset the pads are input enabled, output disabled. Disconnect the
pad's input and output buffers, as well as any pullups. This can reduce
input leakage current.

Signed-off-by: Andrew Featherstone <andrew.featherstone@gmail.com>

ajf58 committed:
drivers: gpio_rpi_pico: Disable inputs when not in use

Reorder gpio_rpi_configure to disable input buffers when not in use.
gpio_rpi_get_config can then determine whether a pin is configured as an
input without requiring additional state variables, as well as reducing
input leakage current.

Signed-off-by: Andrew Featherstone <andrew.featherstone@gmail.com>

soburi commented on drivers/gpio/gpio_rpi_pico.c:
```diff
@@ -50,6 +58,11 @@ static int gpio_rpi_configure(const struct device *dev,
+		gpio_set_dir(pin, GPIO_IN);
+	} else {
+		gpio_set_input_enabled(pin, false);
+	}
```
Insert blank.

soburi replied:
It's a matter of habit. When you look at other code, it's obvious that line breaks are often inserted after `}` and you probably know that too.
The role of reviews is to check for areas that aren't flagged by automated checks.

If I try to give a clear reason, it forms a single semantic unit as a block and has a relationship of parallelism, contraposition, etc. with the next phrase.

soburi commented on drivers/gpio/gpio_rpi_pico.c:
```diff
@@ -77,13 +90,40 @@ static int gpio_rpi_configure(const struct device *dev,
+		if (data->single_ended_mask & BIT(pin)) {
+			*flags |= data->open_drain_mask & BIT(pin) ? GPIO_OPEN_DRAIN : GPIO_PUSH_PULL;
+		}
+	}
```
Insert blank.

ThreeEights commented on drivers/gpio/gpio_rpi_pico.c:
```diff
@@ -50,6 +58,11 @@ static int gpio_rpi_configure(const struct device *dev,
+	if (flags & GPIO_INPUT) {
+		gpio_set_dir(pin, GPIO_IN);
+	} else {
+		gpio_set_input_enabled(pin, false);
```
Given that the RP2xxx series is documented as supporting reading from input pins at any time, does it make sense to disable the pad here?  It seems to me that the pad should only be disabled if the application explicitly requests.  (Of course, that leaves the question of should there be a Zephyr way to explicitly disable an input ...)

ThreeEights commented on drivers/gpio/gpio_rpi_pico.c:
```diff
@@ -77,13 +90,40 @@ static int gpio_rpi_configure(const struct device *dev,
+	if (gpio_get_dir(pin)) {
+		*flags |= gpio_get_out_level(pin) ? GPIO_OUTPUT_HIGH : GPIO_OUTPUT_LOW;
+		if (data->single_ended_mask & BIT(pin)) {
+			*flags |= data->open_drain_mask & BIT(pin) ? GPIO_OPEN_DRAIN : GPIO_PUSH_PULL;
```
Don't you hate that?  :-)

ajf58 reviewed (commented):
This push:
1. Rebases onto a newer revision of `main`.
2. Applies whitespaces requested in the review (although I would like to understand their rationale).
3. Corrects compliance issues that came up under CI.

soburi commented on drivers/gpio/gpio_rpi_pico.c:
```diff
@@ -182,15 +225,61 @@ static int gpio_rpi_manage_callback(const struct device *dev,
+		if (inputs && (pads_bank0_hw->io[pin] & PADS_BANK0_GPIO0_IE_BITS)) {
+			*inputs |= BIT(pin);
+		}
+	}
+	if (inputs) {
+		*inputs &= map;
+	}
```
Here too, add a blank after `}`.

ajf58 replied:
No.

256-258 is related to 259-261, See also 249-254, 107-112. These are logically grouped together, and that's represented by the code not have vertical whitespace between them.

ajf58 commented:
At this point I think this PR it GTG.
@soburi , please take some time to read the reviewer guidelines at https://github.com/zephyrproject-rtos/zephyr/blob/main/doc/contribute/reviewer_expectations.rst?plain=1#L39-L55 . I feel those 3 points are applicable.

soburi commented:
> At this point I think this PR it GTG. @soburi , please take some time to read the reviewer guidelines at https://github.com/zephyrproject-rtos/zephyr/blob/main/doc/contribute/reviewer_expectations.rst?plain=1#L39-L55 . I feel those 3 points are applicable.

I believe there is no problem based on the following wording in the document you showed.

> Style changes that the reviewer disagrees with but that are not documented as part of the project can be pointed out as non-blocking,
